stored_requests/backends/http_fetcher/fetcher.go

```diff
@@ -80,8 +82,44 @@ func (fetcher *HttpFetcher) FetchRequests(ctx context.Context, requestIDs []stri
 	return
 }
 
-func (fetcher *HttpFetcher) FetchCategories(primaryAdServer, publisherId, iabCategory string) (string, error) {
-	return "", nil
+func (fetcher *HttpFetcher) FetchCategories(ctx context.Context, primaryAdServer, publisherId, iabCategory string) (string, error) {
+	if fetcher.Categories == nil {
+		fetcher.Categories = make(map[string]map[string]file_fetcher.Category)
+	}
+
+	if publisherId == "" {
+		publisherId = primaryAdServer
+	}
+
+	dataName := fmt.Sprintf("%s_%s", primaryAdServer, publisherId)
+	if data, ok := fetcher.Categories[dataName]; ok {
+		return data[iabCategory].Id, nil
+	}
+
+	url := fmt.Sprintf("%s/%s/%s.json", strings.Replace(fetcher.Endpoint, "?", "", -1), primaryAdServer, publisherId)
```

Comment:
Probably just removing the ? isn't going to work if the fetcher.Endpoint contains query params in the URL, we should just have this function throw an appropriate error if the endpoint has a ? here (instead of trying to handle that case).

Reply:
I made it because in `NewFetcher` function there is this code: 

```
func NewFetcher(client *http.Client, endpoint string) *HttpFetcher {
	// Do some work up-front to figure out if the (configurable) endpoint has a query string or not.
	// When we build requests, we'll either want to add `?request-ids=...&imp-ids=...` _or_
	// `&request-ids=...&imp-ids=...`, depending.
	urlPrefix := endpoint
	if strings.Contains(endpoint, "?") {
		urlPrefix = urlPrefix + "&"
	} else {
		urlPrefix = urlPrefix + "?"
	}

	...

	return &HttpFetcher{
		client:   client,
		Endpoint: urlPrefix,
	}
} 
```
because of this there is always `?` at the end of url. 
In my understanding in case of Categories fetcher we expect to have url without parameters, so I have to handle this explicitly. 
Do you think I need to refactor existing functionality?

Reply:
Oh ok I see why that's there then. I guess the difference is category fetcher is expecting it's variable parts to be in the URL path where as other fetchers are passing them in URL query params. Maybe this is fine then, but please add a comment to the effect that you are stripping away the question mark that gets added in NewFetcher code. Also we're assuming endpoints used for category mappings aren't going to contain query params, which maybe is ok since they'll probably just be statically hosted files.

Reply:
Comment added.